diff --git a/configure b/configure
index 00b5d9795e..b25e5963ef 100755
--- a/configure
+++ b/configure
@@ -222,6 +222,7 @@ External library support:
                            needed for subtitles and ass filter [no]
   --enable-libbluray       enable BluRay reading using libbluray [no]
   --enable-libbs2b         enable bs2b DSP library [no]
+  --enable-libcbmx         enable C-BMX library [no]
   --enable-libcaca         enable textual display using libcaca [no]
   --enable-libcelt         enable CELT decoding via libcelt [no]
   --enable-libcdio         enable audio CD grabbing with libcdio [no]
@@ -1388,6 +1389,7 @@ test_pkg_config(){
     test_cmd $pkg_config --exists --print-errors $pkg_version || return
     pkg_cflags=$($pkg_config --cflags $pkg_config_flags $pkg)
     pkg_libs=$($pkg_config --libs $pkg_config_flags $pkg)
+
     check_func_headers "$headers" "$funcs" $pkg_cflags $pkg_libs "$@" &&
         enable $name &&
         set_sanitized "${name}_cflags"    $pkg_cflags &&
@@ -1710,6 +1712,7 @@ EXTERNAL_LIBRARY_LIST="
     libass
     libbluray
     libbs2b
+    libcbmx
     libcaca
     libcelt
     libcodec2
@@ -3164,6 +3167,9 @@ asf_stream_muxer_select="asf_muxer"
 av1_demuxer_select="av1_frame_merge_bsf av1_parser"
 avi_demuxer_select="riffdec exif"
 avi_muxer_select="riffenc"
+cbmx_muxer_select="bmx_muxer"
+cbmx_muxer_deps="libcbmx"
+bmx_muxer_deps="libcbmx"
 caf_demuxer_select="iso_media"
 caf_muxer_select="iso_media"
 dash_muxer_select="mp4_muxer"
@@ -6049,6 +6055,14 @@ for func in $COMPLEX_FUNCS; do
     eval check_complexfunc $func \${${func}_args:-1}
 done
 
+add_cflags -I/usr/local/include/bmx-0.1
+add_cflags -I/usr/local/include/libMXF++-1.0
+add_cflags -I/usr/local/include/libMXF-1.0
+add_cflags -fPIC
+add_ldflags -lbmx-0.1
+add_ldflags -lMXF++-1.0
+add_ldflags -lMXF-1.0
+
 # these are off by default, so fail if requested and not available
 enabled avisynth          && require_headers "avisynth/avisynth_c.h"
 enabled cuda_nvcc         && { check_nvcc cuda_nvcc || die "ERROR: failed checking for nvcc."; }
@@ -6066,6 +6080,8 @@ enabled libiec61883       && require libiec61883 libiec61883/iec61883.h iec61883
 enabled libass            && require_pkg_config libass libass ass/ass.h ass_library_init
 enabled libbluray         && require_pkg_config libbluray libbluray libbluray/bluray.h bd_open
 enabled libbs2b           && require_pkg_config libbs2b libbs2b bs2b.h bs2b_open
+enabled libcbmx           && { check_pkg_config libcbmx libcbmx "cbmx/version.h" get_bmx_library_name ||
+                               require libcbmx "cbmx/version.h" get_bmx_library_name -lcbmx -lbmx-0.1 -lMXF++-1.0 -lMXF-1.0 -lstdc++; }
 enabled libcelt           && require libcelt celt/celt.h celt_decode -lcelt0 &&
                              { check_lib libcelt celt/celt.h celt_decoder_create_custom -lcelt0 ||
                                die "ERROR: libcelt must be installed and version must be >= 0.11.0."; }
diff --git a/libavformat/Makefile b/libavformat/Makefile
index e4d997c4a0..ef05e92fdc 100644
--- a/libavformat/Makefile
+++ b/libavformat/Makefile
@@ -124,6 +124,7 @@ OBJS-$(CONFIG_BINTEXT_DEMUXER)           += bintext.o sauce.o
 OBJS-$(CONFIG_BIT_DEMUXER)               += bit.o
 OBJS-$(CONFIG_BIT_MUXER)                 += bit.o
 OBJS-$(CONFIG_BMV_DEMUXER)               += bmv.o
+OBJS-$(CONFIG_BMX_MUXER)                 += bmxenc.o
 OBJS-$(CONFIG_BOA_DEMUXER)               += boadec.o
 OBJS-$(CONFIG_BFSTM_DEMUXER)             += brstm.o
 OBJS-$(CONFIG_BRSTM_DEMUXER)             += brstm.o
diff --git a/libavformat/allformats.c b/libavformat/allformats.c
index 498077e1de..3a1782fc91 100644
--- a/libavformat/allformats.c
+++ b/libavformat/allformats.c
@@ -85,6 +85,7 @@ extern AVInputFormat  ff_bink_demuxer;
 extern AVInputFormat  ff_bit_demuxer;
 extern AVOutputFormat ff_bit_muxer;
 extern AVInputFormat  ff_bmv_demuxer;
+extern AVOutputFormat ff_bmx_muxer;
 extern AVInputFormat  ff_bfstm_demuxer;
 extern AVInputFormat  ff_brstm_demuxer;
 extern AVInputFormat  ff_boa_demuxer;
diff --git a/libavformat/bmxenc.c b/libavformat/bmxenc.c
index 6e4c5f8fa3..89e1af42a0 100644
--- a/libavformat/bmxenc.c
+++ b/libavformat/bmxenc.c
@@ -24,7 +24,7 @@ static int bmx_writer_init(AVFormatContext *s)
     config.shim_name = bmx->shim_name;
     config.frame_rate_num = st->time_base.den;
     config.frame_rate_den = st->time_base.num;
-    config.part = bmx->part;
+    config.partition_size_in_frames = bmx->partition_size_in_frames;
     config.aes3 = bmx->aes3;
     config.kag_512 = bmx->kag_512;
     config.single_pass = bmx->single_pass;
@@ -124,7 +124,7 @@ static const AVOption options[] = {
     { "aes3", "Enable wrapping of audio in AES-3 pair", OFFSET(aes3), AV_OPT_TYPE_FLAGS, { .i64 = 0 }, 0, UINT_MAX, E},
     { "kag_512", "Enable KAG Size 512", OFFSET(kag_512), AV_OPT_TYPE_FLAGS, { .i64 = 0 }, 0, UINT_MAX, E},
     { "single_pass", "Enable single pass writing", OFFSET(single_pass), AV_OPT_TYPE_FLAGS, { .i64 = 0 }, 0, UINT_MAX, E},
-    { "part", "Video essence partition interval in frames", OFFSET(part), AV_OPT_TYPE_DURATION, {.i64 = -1 }, -1, INT_MAX, E},
+    { "part", "Video essence partition interval in frames", OFFSET(partition_size_in_frames), AV_OPT_TYPE_INT, {.i64 = -1 }, -1, INT_MAX, E},
     { NULL },
 };
 
diff --git a/libavformat/bmxenc.h b/libavformat/bmxenc.h
index 7beeed5067..41d06ffd48 100644
--- a/libavformat/bmxenc.h
+++ b/libavformat/bmxenc.h
@@ -9,7 +9,7 @@ typedef struct BMXContext {
     const AVClass* class;  /* Class for private options. */
     ClipType clip_type;
     ShimName shim_name;
-    float part;
+    int partition_size_in_frames;
     int aes3;
     int kag_512;
     int single_pass;
